docs/docs/specs/v0/composer-2.md

```diff
@@ -0,0 +1,389 @@
+
+## Example Structure
+
+Below is a diagram illustrating how:
+
+- a full application is built up to handle a `Stack` of `Composer` and `Image`
+- `Surfaces` are the only element of abstraction necessary to both compose the entire application chrome and implement the abstract portion of `Stack`.
+
+![Composer Structural Diagram](./composer.drawio.svg)
+
+## Stacks
+
+- Stacks are just components which rely on the same `<Surface />` concept to delegate the concrete rendering of stack elements to components provided by plugins.
+
+## Routing
+
+- Routing (the main router and top level routes) are declared outside (around) the `<App />` component and allow the `App` to pass route information as state in the total `AppState` context for all descendants.
+- Nested components inside `<App />` are equally free to use `<Route />` elements if necessary as well.
+- Nested components are encouraged to perform navigation by submitting `Actions` as opposed to calling `navigate()` directly, although situations like `<Button to="/some/route"/>` should also be supported.
+
+## Application State and Actions
+
+- The App provides a global `AppState` (in the `AppContext`) to all nested components which is composed of some concrete and other dynamic sections to be filled by plugins
+- State is updated via a `dispatch(...actions)` API in the `AppContext` which enables Plugins to subscribe to each other's actions and update state accordingly
+- Plugins (and their provided Components) are free to extend/add more fields to the `AppState` as necessary, when it is desirable to share certain state application-wide.
+- State can be composed of `@preact/signals` (or similar reactive state solution) and can contain any other objects including `ECHO` objects.
+- When state is:
+  - local, ephemeral, per-session: it is managed by signals (and backed by local or session storage as appropriate). i.e. the selection state in a List or Tree is ephemeral and not shared between windows or devices, but may be shared among components and plugins in the application.
+  - peer-to-peer, permanent: it is managed by ECHO objects.
+- `AppState` includes current route and params in convenient form.
+- Actions are handled by functions that resemble `reducers` who should return new state reflecting the action committing, and possibly a list of side-effects which may result in more actions or state updates at a later time. (i.e. performing writes to echo objects in a batch is a side-effect, invoking a github API is a side-effect, etc. whereas the indicator that the app is loading something is a purely new state returned from the reducer)
+
+## Plugins, Components, Actions, and other `provides`
+
+- The application framework consumes a list of plugins (statically compiled-in for now).
+- Plugins can provide Components in response to a specific data context: i.e `plugin.provides.component(data: any): React.FC`
```

Comment:
I think this needs to be a predicate the engine can use to determine if a component can be rendered:
```tsx
plugin.provides.component(data: any): ReactNode | null
```

Similar to Fluent Blocks’ `renderIf…(data:any)`

Reply:
Also, the predicate signature probably needs at least two args, either nested in `data` or what have you: the entity the engine is looking for a plugin to render, and maybe the global state as well.

Reply:
exactly, perhaps the interface needs more massage for sure, but if you squint at what you wrote above, that signature is remarkably like a component (FC) just that you called the props `data`. ;) Basically yes I agree, we just have to rationalize with the idea that FCs are just regular functions that can already conditionally return ReactNode, except that since they can also have side effects, perhaps we should wrap them in one more level of function, such as: 
```tsx
plugin.provides.component(data: any) => (props: any) => ReactNode
```

Reply:
perhaps all the components are expected to take a specific props type such as
```
type ComponentProps<TData> = {
  data: TData;
  surfaceContext: any
  ...
}
```
but also, such components are welcome to just use hooks like `useAppState` and/or `useSurfaceContext` to figure out where they are etc. all good options

Reply:
> components are welcome to just use hooks

A hook might not work well for a synchronous predicate, which I imagine the engine will need in order to quickly find the component of best fit for a surface-datum-appstate… the component should use the hooks to stay up-to-date, but otherwise I think it needs to resolve in something-or-null quickly for good UX.

Reply:
Not only that, the engine will probably need a way of re-evaluating which components get to occupy which surfaces… in React, this probably means memoizing on the set of dependencies for the range of possible components’ predicates… something to think about.